File: 백준/Gold/2023. 신기한 소수/신기한 소수.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio                        \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL)
#define ll long long

bool isPrime[10005];
void precalc()
{
    fill(isPrime, isPrime + 10005, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; i * i <= 10000; i++)
    {
        if (isPrime[i])
            for (int j = i * i; j <= 10000; j += i)
            {
                isPrime[j] = false;
            }
    }
}
bool prime(int x)
{
    if (x == 1)
        return false;
    for (int i = 2; i * i <= x; i++)
    {
        if (isPrime[i] && (x % i == 0))
            return false;
    }
    return true;
}

void solve(int cur, int idx)
{
    if (idx == 0)
    {
        cout << cur << '\n';
        return;
    }
    for (int i = 1; i <= 9; i++)
    {
        int nxt = cur * 10 + i;
        if (prime(nxt))
            solve(nxt, idx - 1);
    }
}
int main()
{
    fastio;
    int n;
    cin >> n;
    precalc();
    solve(0, n);
    return 0;
}

File: 백준/Gold/2023. 신기한 소수/신기한 소수.py
import sys

def solve():
    n:int = int(1e5)
    N:int = int(sys.stdin.readline().strip())
    
    def define():
    
        ret = [True for _ in range(n+1)]
        ret[0] = ret[1] = False
        
        for i in range(n + 1):
            if not ret[i]:
                continue
            for j in range(i * i, n + 1, i):
                ret[j] = False
        '''
        for i, state in enumerate(ret):
            print(i, state)
        '''
        res = []
        for i, state in enumerate(ret):
            if state:
                res.append(i)
                
        return res
    
    arr = define()
    ret = []

    def prime(cur: int) -> bool:
        for i in arr:
            if i == cur:
                continue
            if cur % i == 0 or cur == 0 or cur == 1:
                return False
        return True
        
    
    def recursion(idx: int, now: int):
        if idx == N:
            ret.append(now)
            return
        
        if idx == 0:
            start = 1
        else:    
            start = 0
            
        for i in range(start, 10):
            nxt = now * 10 + i
            if prime(nxt):
                recursion(idx+1, nxt)
    recursion(0, 0)
    sorted(ret)
    print(*ret, sep='\n')
    
solve()